Core/HLE/sceNetAdhoc.cpp

```diff
@@ -113,8 +120,457 @@ void __NetAdhocShutdown() {
 	}
 }
 
+static void __AdhocctlNotify(u64 userdata, int cyclesLate) {
+	SceUID threadID = userdata >> 32;
+	int uid = (int)(userdata & 0xFFFFFFFF);
+
+	s64 result = 0;
+	u32 error = 0;
+
+	SceUID waitID = __KernelGetWaitID(threadID, WAITTYPE_NET, error);
+	if (waitID == 0 || error != 0)
+		return; // FIXME: Is it safe to exit here like this without re-scheduling the event? Will this event be triggered again? What will happen to the result i might want to change if exited here?
+
+	int waitVal = __KernelGetWaitValue(threadID, error); // FIXME: Is this value going to be a valid value if waitID == 0? or it's a value belonged to other event?
+	if (adhocctlState != waitVal && error == 0) {
+		// Detecting Adhocctl Initialization using waitVal < 0
+		if (waitVal >= 0 || (waitVal < 0 && (g_Config.bEnableWlan && !networkInited))) {
+			u64 now = (u64)(real_time_now() * 1000.0);
+			if (now - adhocctlStartTime <= adhocDefaultTimeout) {
+				// Try again in another 0.5ms until state matched or timedout.
+				CoreTiming::ScheduleEvent(usToCycles(500) - cyclesLate, adhocctlNotifyEvent, userdata);
+				return;
+			}
+			else
+				result = 0; // ERROR_NET_ADHOCCTL_BUSY
+		}
+		else
+			result = 0; // Faking successfully connected to adhoc server
+	}
+
+	//HLEKernel::ResumeFromWait(threadID, WAITTYPE_NET, uid, result); // FIXME: This won't do anything if waitID == 0, not sure what kind of value returned from the HLE which i might want to change here.
```

Comment:
Hm, it seems like it'd be better to put the desired adhocctlState in waitID, and keep the waitVal for the return value.  Might be tricky for save states.

You could also keep an unordered_map of thread id -> return value, and store the value there when scheduling the wait.

-[Unknown]

Reply:
Well i'm trying to use something unique as waitID, since windows socket is kinda unique and there is only one socket that communicate with adhoc server so i'm using it, it's also easier to use compared to create yet another global vector/map :D